testfixtures/comparison.py

```diff
@@ -384,7 +419,45 @@ def compare(x, y, **kw):
 
     raise AssertionError(message)
 
-    
+
+def _get_expected_actual(x, y, expected, actual):
+    """
+    Helper method for compare() to work out if it is being invoked with unnamed
+    positional arguments x and y, or if expected/actual kwargs have been
+    specified allowing us to provide more helpful error messages.
+
+    Returns a tuple (x, y, actual_expected) where actual_expected is a bool
+    indicating if x and y are named the actual and expected values,
+    respectively.
+
+    not_there is a singleton value used as the default for all four arguments
+    to compare() to differentiate them from an explicitly passed None value.
+
+    Here are the possible inputs and corresponding outputs:
+        (x=a, y=b, expected=not_there, actual=not_there) -> (a, b, False)
+        (x=not_there, y=not_there, expected=a, actual=b) -> (a, b, True)
+        (x=a, y=not_there, expected=a, actual=not_there) -> (b, a, True)
+        (x=a, y=not_there, expected=not_there, actual=b) -> (a, b, True)
+        Anything else -> ValueError
+    """
+    count = sum((0 if x is not_there else 1 for x in (x, y, expected, actual)))
+    if count != 2:
+        raise ValueError('Exactly 2 of the arguments (x, y, expected, actual) '
+                         'must be specified.')
+
+    if x is not_there and y is not not_there:
+        raise ValueError('Must specify x and y together, '
+                         'or x with one of expected/actual.')
+
+    if expected is not_there and actual is not_there:
+        return x, y, False
+    if x is not_there and y is not_there:
+        return expected, actual, True
+    if actual is not_there:
+        return expected, x, True
+    return x, actual, True
```

Comment:
DEFINITELY not a fan of this huge extra method.


I was envisaging more:

```python
def compare(*args, **kwargs):
    ...
    if args:
        x, y = args
   else:
        x = kwargs.pop('expected')
        y = kwargs.pop('actual')
```

Not sure anything more fancy is required.

Definitely not keen to support anything other than `compare(a, b, ...)` or `compare(expected=a, actual=b)`.